### PR TITLE
Add registry auth tests, fix auth short-circuit

### LINKED ISSUE #3129: registry Authorize check returns early
AccessController#Authorized loops over requested access records:
```
	for _, access := range accessRecords {
		log.Debugf("OpenShift auth: checking for access to %s:%s:%s", access.Resource.Type, access.Resource.Name, access.Action)

		switch access.Resource.Type {
		case "repository":
                        ...
			return WithUserClient(ctx, client), nil
		case "admin":
			switch access.Action {
			case "prune":
				...
				return ctx, nil
			default:
				challenge.err = fmt.Errorf("Unknown action: %s", access.Action)
				return nil, challenge
			}
		}
	}
```

If the first accessRecord requested is allowed, the method returns and doesn't check the rest.

Also, if an unknown resource type is encountered, it is ignored, rather than failing

### LINKED ISSUE #3179: Misleading error message when registry is misconfigured
If I happen to mess up something about my registry credentials such that it can't report the build, the end-user error can be misleading:

```
builder.go:70] Build error: Failed to push image: Error pushing to registry: Authentication is required.
```

In my reproducer below, the problem is that the registry is misconfigured or for some other reason can't reach the master (could be DNS/network broken...). It's not something that the person doing a build can do anything about. The message to the user should differentiate between "you don't have access to push to that" (which is the most likely interpretation of the current message) versus "the registry wasn't able to reach the master to report this image".

```
# systemctl start openshift
# cp /openshift.local.config/master/openshift-registry.kubeconfig /tmp
# vi /tmp/openshift-registry.kubeconfig 
    (modify server to point at https://bork.local:8443 or something else broken)
# oadm registry --create --credentials=/tmp/openshift-registry.kubeconfig
# oc process -f examples/sample-app/application-template-dockerbuild.json | oc create -f -
# oc build-logs --follow ruby-sample-build-1
[...]
Successfully built 67ecd0a5afa9
I0615 12:49:49.006070       1 docker.go:87] Pushing 172.30.30.93:5000/default/origin-ruby-sample image ...
E0615 12:49:51.678455       1 dockerutil.go:50] push for image 172.30.30.93:5000/default/origin-ruby-sample failed, will retry in 10s ...
E0615 12:50:03.979110       1 dockerutil.go:50] push for image 172.30.30.93:5000/default/origin-ruby-sample failed, will retry in 10s ...
F0615 12:50:20.365042       1 builder.go:70] Build error: Failed to push image: Error pushing to registry: Authentication is required.

#  oc logs docker-registry-1-5i000
time="2015-06-15T12:50:13Z" level=info msg="response completed" http.request.host="172.30.30.93:5000" http.request.id=a881107e-14bd-48fc-b6ed-68f61f56aa8d http.request.method=GET http.request.remoteaddr="172.17.42.1:39936" http.request.uri="/v2/" http.request.useragent="docker/1.6.0 go/go1.4.2 kernel/3.18.3-201.fc21.x86_64 os/linux arch/amd64" http.response.contenttype="application/json; charset=utf-8" http.response.duration=2.646224ms http.response.written=0 instance.id=4cd34ebc-f70f-4318-9905-0e2e60704305 
172.17.42.1 - - [15/Jun/2015:12:50:13 +0000] "GET /v2/ HTTP/1.1" 401 114 "" "docker/1.6.0 go/go1.4.2 kernel/3.18.3-201.fc21.x86_64 os/linux arch/amd64"
time="2015-06-15T12:50:16Z" level=debug msg="authorizing request" http.request.host="172.30.30.93:5000" http.request.id=14ccae83-82f5-4b4d-b219-40225de003e5 http.request.method=POST http.request.remoteaddr="172.17.42.1:39942" http.request.uri="/v2/default/origin-ruby-sample/blobs/uploads/" http.request.useragent="docker/1.6.0 go/go1.4.2 kernel/3.18.3-201.fc21.x86_64 os/linux arch/amd64" instance.id=4cd34ebc-f70f-4318-9905-0e2e60704305 vars.name="default/origin-ruby-sample" 
time="2015-06-15T12:50:16Z" level=debug msg="OpenShift auth: checking for access to repository:default/origin-ruby-sample:pull" 
time="2015-06-15T12:50:16Z" level=error msg="OpenShift client error: Post https://bork.local:8443/oapi/v1/namespaces/default/subjectaccessreviews: dial tcp: lookup bork.local: no such host" 
time="2015-06-15T12:50:16Z" level=error msg="error authorizing context: openshift access denied" http.request.host="172.30.30.93:5000" http.request.id=14ccae83-82f5-4b4d-b219-40225de003e5 http.request.method=POST http.request.remoteaddr="172.17.42.1:39942" http.request.uri="/v2/default/origin-ruby-sample/blobs/uploads/" http.request.useragent="docker/1.6.0 go/go1.4.2 kernel/3.18.3-201.fc21.x86_64 os/linux arch/amd64" instance.id=4cd34ebc-f70f-4318-9905-0e2e60704305 vars.name="default/origin-ruby-sample" 
```

It appears the registry interprets *any* error returned from the master as a permission error, while there are a large number of other possible error conditions.

### PR DESCRIPTION
**Pull request by liggitt**
Fixes #3129
Fixes #3179

### CODE REVIEW

**Comment by liggitt** [comment]
@ncdc still testing, but PTAL

File: pkg/dockerregistry/server/auth.go [hunk 1]
```diff
@@ -151,15 +137,15 @@ func (ac *AccessController) Authorized(ctx context.Context, accessRecords ...reg
...
 				return nil, challenge
 			}
 
-			if err := verifyImageStreamAccess(repoParts[0], repoParts[1], verb, client); err != nil {
+			if err := verifyImageStreamAccess(imageStreamNS, imageStreamName, verb, client); err != nil {
 				challenge.err = err
 				return nil, challenge
 			}
 
-			return WithUserClient(ctx, client), nil
```
Comment (liggitt) [inline pkg/dockerregistry/server/auth.go]: moved this to the end of the function, to continue looping through all access records

File: pkg/dockerregistry/server/auth.go [hunk 2]
```diff
@@ -167,16 +153,57 @@ func (ac *AccessController) Authorized(ctx context.Context, accessRecords ...reg
 					challenge.err = err
 					return nil, challenge
 				}
-
-				return ctx, nil
```
Comment (liggitt) [inline pkg/dockerregistry/server/auth.go]: removed in lieu of the return at the bottom of the function. should we have been returning `WithUserClient(ctx, client)` here?

File: pkg/dockerregistry/server/auth.go [hunk 3]
```diff
@@ -167,16 +153,57 @@ func (ac *AccessController) Authorized(ctx context.Context, accessRecords ...reg
...
 			default:
 				challenge.err = fmt.Errorf("Unknown action: %s", access.Action)
+				log.Error(challenge.err)
 				return nil, challenge
 			}
+		default:
```
Comment (liggitt) [inline pkg/dockerregistry/server/auth.go]: should we default to disallowing auth to resource types we don't expect?

**Comment by liggitt** [comment]
[test]

File: pkg/dockerregistry/server/auth.go (hunk 2 above)
Comment (ncdc) [inline pkg/dockerregistry/server/auth.go]: Adding the user client to the context is only necessary if any of the code down the line needs to talk to the OpenShift master using the user's supplied token.

Comment (liggitt) [inline pkg/dockerregistry/server/auth.go]: does it cause problems if we do it in all cases? not sure what to do if a mix of repository/push and admin/prune access records are requested

Comment (ncdc) [inline pkg/dockerregistry/server/auth.go]: It doesn't cause any problems, but the way the registry is coded now, you'll either get pull, pull+push, or prune.

Also, the only place the user client is used any more is when a push happens and the target image stream doesn't exist, the registry uses the user's credentials to try to create the image stream.

File: pkg/dockerregistry/server/auth.go (hunk 3 above)
Comment (ncdc) [inline pkg/dockerregistry/server/auth.go]: Probably

**Comment by liggitt** [comment]
@ncdc added check for "*" with message about using `oadm prune images`, PTAL

File: pkg/dockerregistry/server/auth.go [hunk 4]
```diff
@@ -149,17 +138,20 @@ func (ac *AccessController) Authorized(ctx context.Context, accessRecords ...reg
 				verb = "update"
 			case "pull":
 				verb = "get"
+			case "*":
+				challenge.err = fmt.Errorf("Direct deletion of images is not supported by this registry. Use 'oadm prune images' instead.")
```
Comment (ncdc) [inline pkg/dockerregistry/server/auth.go]: I assume that most normal users won't be allowed to run prune.

Comment (liggitt) [inline pkg/dockerregistry/server/auth.go]: just leave it at "direct delete not allowed"?

Comment (ncdc) [inline pkg/dockerregistry/server/auth.go]: I think that's better

**Comment by smarterclayton** [comment]
[test]

**Comment by liggitt** [comment]
ready for review

**Comment by liggitt** [comment]
```
error: timed out waiting for the condition
!!! Error in hack/test-cmd.sh:689
  'oc delete all --all' exited with status 1
Call stack:
  1: hack/test-cmd.sh:689 main(...)
```
Re[test]

**Comment by smarterclayton** [comment]
Changes look ok to me, Andy gets final

**Comment by ncdc** [comment]
Resuming my review now

**Comment by ncdc** [comment]
LGTM

**Comment by liggitt** [comment]
[merge]

**Comment by liggitt** [comment]
```
--- FAIL: TestImage (1.13s)
	client_test.go:117: error retrieving image https://registry-1.docker.io/v1/images/c1a0f06934d152934771963773427a57afff28d5890d2e71d3cf1040d9a9f4c6/json: server returned 504
FAIL
coverage: 71.7% of statements
FAIL	github.com/openshift/origin/pkg/dockerregistry	1.373s
```
Though I get clean test runs locally and Jenkins test passed. Hrmmm

**Comment by smarterclayton** [comment]
Can happen

**Comment by liggitt** [comment]
ok, that's testing against dockerhub... just double-checking that wasn't hitting code changed in the PR. re[merge]

**Comment by liggitt** [comment]
this PR is cursed:
```
Cannot find a valid baseurl for repo: base/7/x86_64
Could not retrieve mirrorlist http://mirrorlist.centos.org/?release=7&arch=x86_64&repo=os&infra=stock error was
12: Timeout on http://mirrorlist.centos.org/?release=7&arch=x86_64&repo=os&infra=stock: (28, 'Operation too slow. Less than 1000 bytes/sec transferred the last 30 seconds')
```
re[merge]

**Pull request merged**